src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,986 @@
+                                // before 
+                                // a = (from a in b select a).ToList();
+                                // after 
+                                // a = new List<T>();
+                                // foreach(...)
+                                variable = assignmentExpression.Left;
+                                nodesBefore = new[] { parentStatement.ReplaceNode(invocationExpression, initializer.WithAdditionalAnnotations(Simplifier.Annotation)) };
+                                nodesAfter = new StatementSyntax[] { };
+                            }
+                            else
+                            {
+                                // before 
+                                //  a = (from a in b select a).ToList();
+                                // after 
+                                // var list = new List<T>();
+                                // foreach(...)
+                                // a = list;
+                                variable = SyntaxFactory.IdentifierName(symbolName);
+                                nodesBefore = new[] { CreateLocalDeclarationStatement(symbolName, initializer, generateTypeFromExpression: false) };
+                                nodesAfter = new StatementSyntax[] { parentStatement.ReplaceNode(invocationExpression, variable.WithAdditionalAnnotations(Simplifier.Annotation)) };
+                            }
+
+                            return true;
+                        }
+
+                        break;
+                    case SyntaxKind.ReturnStatement:
+                        // before return (from a in b select a).ToList();
+                        // after var list = new List<T>();
+                        // foreach(...)
+                        // return list;
+                        variable = SyntaxFactory.IdentifierName(symbolName);
+                        nodesBefore = new[] { CreateLocalDeclarationStatement(symbolName, initializer, generateTypeFromExpression: false) };
+                        nodesAfter = new[] { SyntaxFactory.ReturnStatement(variable).WithAdditionalAnnotations(Simplifier.Annotation) };
+                        return true;
+                        // SyntaxKind.Argument:
+                        // SyntaxKind.ArrowExpressionClause is not supported
+                }
+
+                // Will still try to replace with a local function above.
+                nodesBefore = default;
+                nodesAfter = default;
+                variable = default;
+                return false;
+            }
+
+            private LocalDeclarationStatementSyntax CreateLocalDeclarationStatement(
+                SyntaxToken identifier,
+                ExpressionSyntax expression,
+                bool generateTypeFromExpression)
+            {
+                var typeSyntax = generateTypeFromExpression ?
+                     _semanticModel.GetTypeInfo(expression).ConvertedType.GenerateTypeSyntax() :
```

Comment:
>GenerateTypeSyntax [](start = 74, length = 18)

I am tot familiar with this API, what does it do if Anonymous Types are involved?

Reply:
1. It creates 'object' for anonymous types.
2. We generate it for selectClauses only where anonymous types are prohibited

---
In reply to: [178423653](https://github.com/dotnet/roslyn/pull/25362#discussion_r178423653) [](ancestors = 178423653)

Reply:
> 1. It creates 'object' for anonymous types.

Object would probably never work, ```var``` has better chances I think.

---
In reply to: [178424121](https://github.com/dotnet/roslyn/pull/25362#discussion_r178424121) [](ancestors = 178424121,178423653)

Reply:
Object definitely won't have members specific to the original type.

---
In reply to: [178424311](https://github.com/dotnet/roslyn/pull/25362#discussion_r178424311) [](ancestors = 178424311,178424121,178423653)

Reply:
We should not have anonymous types here. Do you mean we should check for anonymous types here once again and generate 'var' in the case?

---
In reply to: [178424327](https://github.com/dotnet/roslyn/pull/25362#discussion_r178424327) [](ancestors = 178424327,178424311,178424121,178423653)

Reply:
> We should not have anonymous types here. Do you mean we should check for anonymous types here once again and generate 'var' in the case?

I don't see what would prevent us from getting here with an anonymous type from line 210:
```
return AddToBlockTop(CreateLocalDeclarationStatement(identifier, selectClause.Expression, generateTypeFromExpression: true), statement);
```

Yes, I suggest to grab converted type and use it only if it doesn't reference anonymous types.

---
In reply to: [178424371](https://github.com/dotnet/roslyn/pull/25362#discussion_r178424371) [](ancestors = 178424371,178424327,178424311,178424121,178423653)

Reply:
Ok, I see that you added a check in TryProcessQueryBody. We should be good then. 

---
In reply to: [178424489](https://github.com/dotnet/roslyn/pull/25362#discussion_r178424489) [](ancestors = 178424489,178424371,178424327,178424311,178424121,178423653)